File: compscores.pl
use strict;

my $CurrBoard = 0;
my $HighestBoard = 0;
my @Data;
my %NameData;
my %TotDup;
my %PairTot;
my $NS = 1;
my $EW = 2;
my %PNums;
my $BText;
my $totalPairs = 0;
my $isMitchell = 0;

my %ScoreName = (
    'TP'   => 'TOTALPOINTS',
    'IMPS' => '    IMPS   ',
    'MP'   => 'MATCHPOINTS',
    );

die "Usage: perl compscores.pl [MP|IMPS|TP] Travs.txt" if (@ARGV < 2); 
my $ScoreType = shift @ARGV;
#print "ScoreType = <$ScoreType>\n";
#exit(0);

while (<>) {
    # format is like 4.00   0.00 9-Giordano-Giordano vs 2-Anderson-Hochens
    if (m|^ *RESULTS OF BOARD (\d+)|) {
	$CurrBoard = $1;
	$HighestBoard = $1;
    }
    elsif (m|^      \s*(\d+) .* (\d+)-(\w+-\w+) vs (\d+)-(\w+-\w+)|) {
	# a minus score for NS, plus for EW
	my $ewscore = 1*"$1";
	my $nsscore = -1*$ewscore;
	SaveScoreData($CurrBoard, $2, $3, $nsscore, $4, $5, $ewscore);
    }
    elsif (m|^\s*(\d+) .* (\d+)-(\w+-\w+) vs (\d+)-(\w+-\w+)|) {
	# a plus score for NS, minus for EW
	my $nsscore = 1*"$1";
	my $ewscore = -1*$nsscore;
	SaveScoreData($CurrBoard, $2, $3, $nsscore, $4, $5, $ewscore);
    }
    elsif (m|^ \s*(PASS) .* (\d+)-(\w+-\w+) vs (\d+)-(\w+-\w+)|) {
	# Passed out, so a zero score for both NS and EW
	SaveScoreData($CurrBoard, $2, $3, 0, $4, $5, 0);
    }
    elsif ($CurrBoard == 0 && m|^\s*(\d+)\s*\d+\.\d+ \s* \d+\.\d+|) {
	$totalPairs++;
    } 

    else {
	if ($CurrBoard>0 && !m|N-S| && !m|SCORES| && !m|-----| && !m|^ $|) {
	    print STDERR "Unmatched line: $_";
	}
    }
}

# show scoredatas by boards read (for debugging)
for (my $bd=1; $bd <= $HighestBoard; $bd++) {
    my $count=0;
    my $str;
    foreach my $pra (sort {$a <=> $b} keys %{$Data[$bd]}) {
	$str .= "$pra,";
	$count++;
    }
    # print STDERR "$bd: $count $str\n";
}
# exit(1);

# now show what we got
for (my $bd=1; $bd <= $HighestBoard; $bd++) {
    #print "-------\nBoard $bd\n";
    #foreach my $pnum (sort {$a <=> $b} keys %{$Data[$bd]}) {
    #	print ("Score for $pnum = $Data[$bd]{$pnum}{SCORE}\n");
    #}
    # for each of NS and EW compute a totalpoints diff array
    foreach my $pra (sort {$a <=> $b} keys %{$Data[$bd]}) {
	foreach my $prb (sort {$a <=> $b} keys %{$Data[$bd]}) {
	    next if ($pra == $prb);
	    next if ($Data[$bd]{$pra}{DIR} ne  $Data[$bd]{$prb}{DIR});
	    my $ascore = $Data[$bd]{$pra}{SCORE};
	    my $bscore = $Data[$bd]{$prb}{SCORE};
	    my $diff = ($ascore - $bscore);
	    my $dup;
	    $dup = $diff if ($ScoreType eq 'TP');
	    $dup = MP($diff) if ($ScoreType eq 'MP');
	    $dup = IMPS($diff) if ($ScoreType eq 'IMPS');
	    # print ("Board $bd, $pra vs. $prb, dup=$dup\n");
	    $Data[$bd]{$pra}{DUPSCORE} += $dup;
	    $TotDup{$pra} += $dup;
	    $PairTot{$pra}{$Data[$bd]{$pra}{OPP}} += $dup;
	}
    }
}
print "----------\nTotals for All Boards\n";
PrintTotals(\%TotDup);
print "\n\n";
for (my $bd=1; $bd <= $HighestBoard; $bd++) {
    PrintOneBoard($bd);
}

# compute top on a board...
my $duptot = 0;
my $pairs = 0;
foreach my $pra (keys %{$Data[1]}) {
    $duptot += $Data[1]{$pra}{DUPSCORE};
    $pairs++;
}
my $boardTop = 2*$duptot / $pairs;

# compute number of boards per round
# try 1 vs. highest
my $bpr = 0;
for (my $bd=1; $bd <= $HighestBoard; $bd++) {
    $bpr++ if ($Data[$bd]{1}{OPP} == $pairs);
}
if ($bpr == 0) {
    for (my $bd=1; $bd <= $HighestBoard; $bd++) {
	$bpr++ if ($Data[$bd]{3}{OPP} == $pairs);
    }
}
print "boards per round = $bpr\n";

# print totals pairs vs pairs
print "\n\n";
foreach my $pra (sort {$a <=> $b} keys %PairTot) {
    print "Pair vs. Pair Totals for $pra-$NameData{$pra}: \n";
    foreach my $prb (reverse sort {$PairTot{$pra}{$a} <=> $PairTot{$pra}{$b}} keys %{$PairTot{$pra}}) {
	if ($ScoreType eq 'MP') {
	    my $total = $PairTot{$pra}{$prb};
	    printf ("%5.1f (%3d%%) %s\n", $total, 100.0*$total/($bpr * $boardTop), "$prb-$NameData{$prb}");
	}
	else {
	    printf ("%5d  %s\n", $PairTot{$pra}{$prb}, "$prb-$NameData{$prb}");
	}
    }
    print "\n";
}

# print distributions of scores
print "\n\nScore Distributions\n";
foreach my $pra (sort {$a <=> $b} keys %PairTot) {
    my %h;
    for (my $bd=1; $bd <= $HighestBoard; $bd++) {
	next if ($Data[$bd]{$pra}{OPP} == 0);
	$h{$Data[$bd]{$pra}{DUPSCORE}}++;
    }
    print "$pra-$NameData{$pra}\n";
    for (my $dupscore = $boardTop; $dupscore >= 0; $dupscore -=0.5) {
	printf "%3.1f: ", $dupscore;
	print '*' x $h{$dupscore};
	print "\n";
    }
    print "\n";
}

sub SaveScoreData() {
    my ($bd, $NSPair, $NSName, $NSScore, $EWPair, $EWName, $EWScore) = @_;
    # print STDERR "SaveScoreData: $bd, $NSPair, $NSName, $NSScore, $EWPair, $EWName, $EWScore\n";
    $NSPair *= 1;
    $EWPair *= 1;
    # if first round shows that NSPair is playing same number EWPair, then it must be a mitchell movement
    $isMitchell = 1 if ($NSPair == $EWPair);
    
    # for mitchell artificially bump the EW Pairs by 1/2 totalPairs
    if ($isMitchell) {
	$EWPair += int(($totalPairs + 1) / 2);
    }
    $Data[$bd]{$NSPair}{DIR} = $NS;
    $Data[$bd]{$NSPair}{SCORE} = $NSScore;
    $Data[$bd]{$EWPair}{DIR} = $EW;
    $Data[$bd]{$EWPair}{SCORE} = $EWScore;
    $Data[$bd]{$NSPair}{OPP} = $EWPair;
    $Data[$bd]{$EWPair}{OPP} = $NSPair;
    $NameData{$NSPair} = $NSName;
    $NameData{$EWPair} = $EWName;
}


sub PrintOneBoard() {
    my $bd = shift;
    my %htmp;

    print "RESULTS OF BOARD $bd\n\n";
    print "   SCORES      $ScoreName{$ScoreType}   NAMES\n";
    print "  N-S   E-W    N-S    E-W\n";

    # build a list of NS Scores for this board
    foreach my $pnum (keys %{$Data[$bd]}) {
	my $score = $Data[$bd]{$pnum}{SCORE};
	push @{$htmp{$score}}, $pnum;
    }
    # print sorted by NS Dup score
    foreach my $score (sort {$b <=> $a} keys %htmp) {
	foreach my $pnum (@{$htmp{$score}}) {
	    next if ($Data[$bd]{$pnum}{DIR} != $NS);
	    my $NSPair = $pnum;
	    my $EWPair = $Data[$bd]{$pnum}{OPP};
	    my $NSScore = $Data[$bd]{$NSPair}{SCORE};
	    my $EWScore = $Data[$bd]{$EWPair}{SCORE};
	    my $NSName = $NameData{$NSPair};
	    my $EWName = $NameData{$EWPair};
	    my $NSDup = $Data[$bd]{$NSPair}{DUPSCORE};
	    my $EWDup = $Data[$bd]{$EWPair}{DUPSCORE};
            my $fmtscore = ($NSScore >= 0 ? sprintf("%5d      ", $NSScore) : sprintf("      %5d", $EWScore));
            my $fmtdup = sprintf(($ScoreType eq 'MP' ? "%5.2f  %5.2f" : "%5d  %5d"), $NSDup, $EWDup);
            printf "%s  %s  $NSPair-$NSName vs $EWPair-$EWName\n", $fmtscore, $fmtdup;
	}
    }
    print "----------------------------\n";
}


sub PrintTotals() {
    my $href = shift;
    my %htmp;
    # print sorted by total score
    foreach my $pnum (keys %{$href}) {
	my $score = ${$href}{$pnum};
	push @{$htmp{$score}}, $pnum;
    }
    foreach my $score (sort {$b <=> $a} keys %htmp) {
	foreach my $pnum (@{$htmp{$score}}) {
	    my $Name = $NameData{$pnum};
	    my $fmtscore = sprintf (($ScoreType eq 'MP'? "%3.2f" : "%5d"), $score);
	    printf "Pair %2d %-20s  %s\n", $pnum, $Name, $fmtscore;
	}
    }
}

sub MP() {
    my $diff = shift;	
    return 1 if ($diff > 0);
    return 0.5	if ($diff == 0);
    return 0;
}

sub IMPS() {
    my $score = shift;
    my $signscore = signof($score);
    $score = abs($score);
    my @steps = (20, 50, 90, 130, 170, 220, 270, 320, 370, 430,
		 500, 600, 750, 900, 1100, 1300, 1500,
		 1750, 2000, 2250, 2500, 3000, 3500, 4000, 99999);
    for (my $i=0; $i<=24; $i++) {
	if ($score < $steps[$i]) {
	    return $i * $signscore;
	}
    }
}

sub signof {
    my $score = shift;
    return ($score < 0 ? -1 : 1);
}

File: compscores.py
import argparse
import re
import sys
import collections
import json

def recursively_default_dict():
    return collections.defaultdict(recursively_default_dict)

#globals
isMitchell= False
args = None
highestBoard = None

PairData = {}
maxResultsPerBoard = 0
numResultsBoard = {}

class ResultObj(object):
    def __init__(self, pnum, isNS, score, oppPair):
        self.pairnum = pnum
        self.isNS = isNS
        self.score = score
        self.opp = oppPair
        
class PairObj(object):
    def __init__(self, pairnum):
        self.results = {}
        self.name = None
        self.pairnum = pairnum
        
def MP(diff):
    if diff > 0:
        return 1
    elif diff == 0:
        return 0.5
    else:
        return 0

def IMPS(diff):
    signdiff = 1 if diff >= 0 else -1
    diff = abs(diff)
    steps = (20, 50, 90, 130, 170, 220, 270, 320, 370, 430,
	     500, 600, 750, 900, 1100, 1300, 1500,
	     1750, 2000, 2250, 2500, 3000, 3500, 4000, 99999)
    for i in range(25):
        if (diff < steps[i]):
            impval = i * signdiff
            # print('diff=%d, imps=%d' % (diff, impval))
            return impval
        
def handleScoreLine(bd, m):
    global isMitchell
    # print('<%s> <%s> %s %s %s %s' % (m[1], m[2], m[3], m[4], m[5], m[6]))
    scoreText = m[1]
    if 'PASS' in scoreText:
        ewscore = 0
        nwscore = 0
    else:
        # separate m[1] into ns and ew scores
        nstext = scoreText[0:4]
        ewtext = scoreText[6:10]
        # print('nstext=<%s>, ewtext=<%s>' % (nstext,ewtext))
        if nstext == '    ':
            ewscore = int(ewtext)
            nsscore = -1 * ewscore
        else:
            nsscore = int(nstext)
            ewscore = -1 * nsscore
    # print('ns=', nsscore, ', ew=', ewscore)
    NSPair = int(m[2])
    NSName = m[3]
    EWPair = int(m[4])
    EWName = m[5]
    # print('handle score line for bd %d, ns %d, ew %d' % (bd, NSPair, EWPair))
    
    # if first round shows that NSPair is playing same number EWPair, then it must be a mitchell movement
    if NSPair == EWPair:
        isMitchell = True
    
    # for mitchell artificially bump the EW Pairs by 1/2 totalPairs
    if isMitchell:
        EWPair += int((args.numPairs + 1) / 2)

    if PairData.get(NSPair) == None:
        PairData[NSPair] = PairObj(NSPair)
    if PairData.get(EWPair) == None:
        PairData[EWPair] = PairObj(EWPair)
    
    PairData[NSPair].results[bd] = ResultObj(NSPair, True, nsscore, EWPair)
    PairData[EWPair].results[bd] = ResultObj(EWPair, False, ewscore, NSPair)
    PairData[NSPair].name = NSName
    PairData[EWPair].name = EWName

def dumpData(obj):
    print(json.dumps(obj, indent=1, default=lambda x: x.__dict__))

#=========================================    
# main program
# argument stuff
parser = argparse.ArgumentParser()
parser.add_argument('-f', '--file', required=True, help='input file')
parser.add_argument('-n', '--numPairs', required=True, type=int, help='number of Pairs')
parser.add_argument('-k', '--scoreKind', default='MP', help='scoring Kind (MP, IMP, TP)')
parser.add_argument('-v', '--verbose', action='store_true', help='increase output verbosity')
args = parser.parse_args()


print('input file is', args.file)
resultsPerBoard = 0
with open(args.file) as fp:
    currBoard = 0
    for line in fp:
        m = re.match('^ *RESULTS OF BOARD (\d+)', line)
        if m:
            currBoard = int(m[1])
            highestBoard = currBoard
            resultsPerBoard = 0
            # print('currBoard =', currBoard)
            continue
        # line which matches traveler entry
        # the first section (with scores or PASS)will be parsed in handleScoreLine
        # if plus for N/S, then m[1] will be non-blank.
        # if plus for E/W, then m[2] will be non-blank.
        m = re.match('^ (.{10}).*(\d+)-(\w+-\w+) vs (\d+)-(\w+-\w+)', line)
        if m:
            handleScoreLine(currBoard, m)
            resultsPerBoard += 1
            numResultsBoard[currBoard] = resultsPerBoard
            maxResultsPerBoard = max(maxResultsPerBoard, resultsPerBoard)
            # print(line.rstrip())
            continue
        # print(line.rstrip())
maxResultsPerBoard = max(maxResultsPerBoard, resultsPerBoard)
print('maxResultsPerBoard = %d' % (maxResultsPerBoard))
if False:
    dumpData(PairData)

# for each of pair on each board compute a "duplicate" score
for bd in range(1, highestBoard+1):
    for pra in sorted(PairData.keys()):
        aresult = PairData[pra].results.get(bd)
        if aresult == None:
            continue
        # now go thru other pairs and compute duplicate score vs. ones sitting same direction
        dupBoard = 0
        for prb in sorted(PairData.keys()):
            # print(bd, pra, prb)
            bresult = PairData[prb].results.get(bd)
            if  bresult == None or pra == prb or aresult.isNS != bresult.isNS:
                 continue
            
            ascore = aresult.score
            bscore = bresult.score
            diff = ascore - bscore
            # print ('%d : %d,%d  --> %d' % (bd, pra, prb, diff))
            if args.scoreKind == 'MP':
                dup = MP(diff)
            elif args.scoreKind == 'TP':
                dup = diff
            elif args.scoreKind == 'IMP':
                dup = IMPS(diff)
            else:
                # todo: handle other scoring types like TP or IMP
                print('%s scoring not supported yet' % (args.scoreKind))
            # $dup = $diff if ($ScoreType eq 'TP');
            dupBoard += dup
        # print('total dup for %d on bd %d is %f' % (pra, bd, dupBoard))
        # MP only, adjust duplicates for boards which have fewer plays
        if args.scoreKind == 'MP':
            numResults = numResultsBoard[bd]
            if numResults != maxResultsPerBoard:
                oldDup = dupBoard
                dupBoard = ((oldDup + 0.5) * (maxResultsPerBoard/numResults) - 0.5)
                print('on bd %d, changed dup score for pair %d from %f to %f' % (bd, pra, oldDup, dupBoard))
        # store in result object
        aresult.dup = dupBoard

# compute total duplicate for each pair, adjusting for noplays
dupTotals = {}
for pnum in sorted(PairData.keys()):
    dupTotal = 0
    numBoards = 0
    for bd in range(1, highestBoard+1):
        # get dup for that board if it exists, else 0
        result = PairData[pnum].results.get(bd)
        if result == None:
            dup = 0
        else:
            dup = result.dup
            numBoards += 1
        dupTotal += dup
    if args.scoreKind == 'MP' and numBoards != highestBoard:
        dupTotal = dupTotal * (highestBoard/numBoards)
    dupTotals[pnum] = dupTotal

def byDupTotal(pnum):
    return dupTotals[pnum]

for pnum in sorted(dupTotals.keys(), key=byDupTotal, reverse=True):
    print('dupTotal for pair %d is %+9.2f' % (pnum, dupTotals[pnum]), end='')
    if args.scoreKind == 'MP':
        maxPossible = highestBoard * (maxResultsPerBoard - 1)
        pct = 100 * dupTotal / maxPossible
        print(' or %.2f%%' % (pct))
    else:
        print()


# show the results on each board
def byDup(result):
    return result.dup

for bd in range(1, highestBoard+1):
    results = {}
    for pnum in sorted(PairData.keys()):
        result = PairData[pnum].results.get(bd)
        if result != None:
            results[pnum] = result
    # show sorted by dup score
    print()
    print('Board', bd)
    for result in (sorted(results.values(), key=byDup, reverse=True)):
        if not result.isNS:
            continue
        if args.scoreKind == 'MP':
            dupStr = '%6.2f' % (result.dup)
        else:
            dupStr = '%+6d' % (result.dup)
        print('%6d  %s  %d-%s vs. %d-%s' % (result.score, dupStr, result.pairnum, PairData[result.pairnum].name, result.opp, PairData[result.opp].name))

#show results sorted by opp
for pnum in sorted(PairData.keys()):
    vsDup = {}
    numResults = {}
    for bd in range(1, highestBoard+1):
        result = PairData[pnum].results.get(bd)
        if result == None:
            continue
        oldDup = vsDup.get(result.opp, 0)
        newDup = oldDup + result.dup
        vsDup[result.opp] = newDup
        oldNum = numResults.get(result.opp, 0)
        newNum = oldNum + 1
        numResults[result.opp] = newNum
    def byVsDup(pnum):
        return vsDup[pnum]
    print()
    print('pair %d-%s' % (pnum, PairData[pnum].name))
    for opp in sorted(vsDup.keys(), key=byVsDup, reverse=True):
        print('vs. %2d-%-25s  %5.2f' % (opp, PairData[opp].name, vsDup[opp]), end='')
        if args.scoreKind == 'MP':
            pct = 100 * vsDup[opp]/(numResults[opp] * (maxResultsPerBoard - 1))
            print(', (%2.0f%%)' % (pct), end='')
        print()
            
sys.exit(1)
